common/v1/payload_json.go

```diff
@@ -0,0 +1,228 @@
+	if disabled, _ := m.Metadata[DisablePayloadShorthandMetadataKey].(bool); disabled {
+		return false, nil, nil
+	}
+	handled, value, err := p.toJSONShorthand()
+	if !handled || err != nil {
+		return handled, nil, err
+	}
+	if m.Indent == "" {
+		b, err = json.Marshal(value)
+	} else {
+		b, err = json.MarshalIndent(value, currIndent, m.Indent)
+	}
+	return true, b, err
+}
+
+// MaybeUnmarshalJSONPB implements
+// [go.temporal.io/api/internal/temporaljsonpb.JSONPBMaybeUnmarshaler.MaybeUnmarshalJSONPB].
+//
+// WARNING: This is internal API and should not be called externally.
+func (p *Payload) MaybeUnmarshalJSONPB(u *jsonpb.Unmarshaler, b []byte) (handled bool, err error) {
+	// If this is nil, ignore (should never be)
+	if p == nil {
+		return false, nil
+	}
+	// If shorthand is disabled, ignore
+	if disabled, _ := u.Metadata[DisablePayloadShorthandMetadataKey].(bool); disabled {
+		return false, nil
+	}
+	// Always considered handled, unmarshaler ignored (unknown fields always
+	// disallowed for non-shorthand payloads at this time)
+	p.fromJSONMaybeShorthand(b)
+	return true, nil
+}
+
+func (p *Payload) toJSONShorthand() (handled bool, value interface{}, err error) {
+	// Only support binary null, plain JSON and proto JSON
+	switch string(p.Metadata["encoding"]) {
+	case "binary/null":
+		// Leave value as nil
+		handled = true
+	case "json/plain":
+		// Must only have this single metadata
+		if len(p.Metadata) != 1 {
+			return false, nil, nil
+		}
+		// We unmarshal because we may have to indent. We let this error fail the
+		// marshaller.
+		handled = true
+		err = json.Unmarshal(p.Data, &value)
+	case "json/protobuf":
+		// Must have the message type and no other metadata
+		msgType := string(p.Metadata["messageType"])
+		if msgType == "" || len(p.Metadata) != 2 {
+			return false, nil, nil
+		}
+		// Since this is a proto object, this must unmarshal to a object. We let
+		// this error fail the marshaller.
+		var valueMap map[string]interface{}
+		handled = true
+		err = json.Unmarshal(p.Data, &valueMap)
+		// Put the message type on the object
+		if valueMap != nil {
+			valueMap["_protoMessageType"] = msgType
+		}
+		value = valueMap
+	}
+	return
+}
+
+func (p *Payload) fromJSONMaybeShorthand(b []byte) {
+	// We need to try to deserialize into the regular payload first. If it works
+	// and there is metadata _and_ data actually present (or null with a null
+	// metadata encoding), we assume it's a non-shorthand payload. If it fails
+	// (which it will if not an object or there is an unknown field or if
+	// 'metadata' is not string + base64 or if 'data' is not base64), we assume
+	// shorthand. We are ok disallowing unknown fields for payloads here even if
+	// the outer unmarshaler allows them.
+	if gogojsonpb.Unmarshal(bytes.NewReader(b), p) == nil && len(p.Metadata) > 0 {
+		// A raw payload must either have data or a binary/null encoding
+		if len(p.Data) > 0 || string(p.Metadata["encoding"]) == "binary/null" {
```

Comment:
> if you reference same string more than once, I would like to see them on top of this file in the const block

In this case, these are basically company-wide constants have have no real home. I can move them to _private_ constants in this file, but it makes little sense to me as they are used by two private methods colocated together.

> Similar to DisablePayloadShorthandMetadataKey but these will be really public (exported).

But that is exported for a reason, there is no reason to export these. I only use them in two private methods and never would a user use them (we already have them exported in the Go SDK). I don't think we need to export more consts out of this mostly-generated package.

> Similar [consts](https://github.com/temporalio/sdk-go/blob/master/converter/metadata.go) in Go SDK could just alias them.

Aliasing was a very poor decision. But assuming you mean just have the Go SDK reference a new set of consts here, what benefit is there? If it's just to satisfy the _idea_ that they're shared string literals, I don't think that has enough benefits practically.

> these consts are very well hardcoded (and immutable) not only in this repo but across many repos in entire org.

Exactly. They might as well be part of our proto definitions, but since they aren't we have string literals in the projects that reference them. It is unfortunate that api-go even exists (the SDK and server should be able to own their own proto generation) and luckily other languages aren't burdened with that. If it was reasonable I would not put this logic here, but since this is our "shared Go place for payloads", we need the specialized JSON conversion here I think.